# For the following questions, work with the bank_reg_training and the bank_reg_test data sets. Use either Python or R to solve each problem.

In [11]:
# Importing Libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
import statsmodels.api as sm
from sklearn.preprocessing import StandardScaler
%matplotlib inline

In [12]:
# Importing Datasets
bank_train = pd.read_csv("/Users/datascience/Desktop/ADS 502 Data Sets/Website Data Sets/bank_reg_training.csv")
bank_test = pd.read_csv("/Users/datascience/Desktop/ADS 502 Data Sets/Website Data Sets/bank_reg_test.csv")

## 34. Use the training set to run a regression predicting Credit Score, based on Debt‐to‐Income Ratio and Request Amount. Obtain a summary of the model. Do both predictors belong in the model?

In [13]:
# Seperate Predictor variables and Target Variables 
X = pd.DataFrame(bank_train[['Debt-to-Income Ratio', 'Request Amount']]) # Predictors
y = pd.DataFrame(bank_train[['Credit Score']]) # Target Variable 


#Add Constant variable to the predictors 
X = sm.add_constant(X)

# Multiple Regression Model using Ordinary Least Squares method
RegModel = sm.OLS(y, X).fit()

# Display Summary Regression Model
RegModel.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:           Credit Score   R-squared:                       0.028
Model:                            OLS   Adj. R-squared:                  0.028
Method:                 Least Squares   F-statistic:                     156.2
Date:                Sun, 07 Nov 2021   Prob (F-statistic):           1.37e-67
Time:                        19:47:16   Log-Likelihood:                -59972.
No. Observations:               10693   AIC:                         1.199e+05
Df Residuals:                   10690   BIC:                         1.200e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
========================================================================================
                           coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------
const                  668.4562      1.336    500.275      0.000     665.837     671.075
Debt-to-Income Ratio   -48.1262      4.785    -10.058      0.000     -57.505     -38.747
Request Amount           0.0011   6.84e-05     15.727      0.000       0.001       0.001
==============================================================================
Omnibus:                     1658.575   Durbin-Watson:                   1.991
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2844.250
Skew:                          -1.021   Prob(JB):                         0.00
Kurtosis:                       4.487   Cond. No.                     1.24e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.24e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Both independent variables are statistically significant(p-values) with a low R-squared value indicates that the independent variables are correlated with the dependent variable. In addition, the adjusted R-squared value is the same value as the R^squared value; this shows that there are no useless predictor variables to the model. Therefore, both predictors belong in the model.

## 35. Validate the model from the previous exercise.

In [14]:
## Verify the Regression Model Results ##

# Seperate Predictor variables and Target Variables 
X_test = pd.DataFrame(bank_test[['Debt-to-Income Ratio', 'Request Amount']]) # Predictors
y_test = pd.DataFrame(bank_test[['Credit Score']]) # Target Variable 


#Add Constant variable to the predictors 
X_test = sm.add_constant(X_test)

# Multiple Regression Model using Ordinary Least Squares method
RegModel_test = sm.OLS(y_test, X_test).fit()

# Display Summary Regression Model
RegModel_test.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:           Credit Score   R-squared:                       0.038
Model:                            OLS   Adj. R-squared:                  0.038
Method:                 Least Squares   F-statistic:                     215.4
Date:                Sun, 07 Nov 2021   Prob (F-statistic):           1.94e-92
Time:                        19:47:16   Log-Likelihood:                -60395.
No. Observations:               10775   AIC:                         1.208e+05
Df Residuals:                   10772   BIC:                         1.208e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
========================================================================================
                           coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------
const                  665.4987      1.328    501.265      0.000     662.896     668.101
Debt-to-Income Ratio   -52.1374      4.826    -10.803      0.000     -61.597     -42.677
Request Amount           0.0013   6.85e-05     19.013      0.000       0.001       0.001
==============================================================================
Omnibus:                     1792.693   Durbin-Watson:                   1.985
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             3194.120
Skew:                          -1.067   Prob(JB):                         0.00
Kurtosis:                       4.600   Cond. No.                     1.25e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.25e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Regression Model using training data and the Regression Model using testing data exhibit similar summary results. Therefore, these results validate the results from the model using the training data.

## 36. Use the regression equation to complete this sentence: “The estimated Credit Score equals....”

The estimated Credit Score equals 668.4562 - 48.1262(Debt-to-Income Ratio) + 0.0011(Request Amount)


## 37. Interpret the coefficient for Debt‐to‐Income Ratio.

The estimated decrease in Credit Score with a Debt-to-Income Ratio is -48.1262, when Request Amount is held constant.  However, generally the debt to income ratio doesn't impact your credit scores, but it is a factor when deciding to approve credit applications.

## 38. Interpret the coefficient for Request Amount.

The estimated increase in Credit Score with a Request Amount is 0.0011, when Request Amount is held constant. This makes sense since Request Amount generally affects whether or not the loan application is approved. Morevoer, applying for loans (i.e., request amount) can temporarily knock a few points off the credit score. Therefore, the low contribution towards credit score is valid. 

## 39. Find and interpret the value of s.

In [15]:
# Calculating Standard Error
s = np.sqrt(RegModel.scale)
print("The standard error of the estime equals to ", s)

The standard error of the estime equals to  66.00195259717187


The size of the model's typical prediction error is about 66.02 (when predicting Credit Score with Debt-to-Income Ratio and Request Amount). 

## 40. Find and interpret R-squared-adjusted. Comment.

The R-squared-adjusted has a value of 0.028. Thus, 2.8% of the variability in Credit Score is accounted by the predictors Debt-to-Income Ratio and Request Amount. Determining Credit Score incorporates many other factors, therefore the regression model's small R-squared-adjusted proportion makes sense.

## 41. Find MAE_Baseline and MAE_Regression, and determine whether the regression model outperformed its baseline model.

In [16]:
from sklearn.metrics import mean_absolute_error #import to use Mean Absolute Error

### MAE Baseline ### (using testing dataset)
y_bar= bank_test['Credit Score'].mean()
ypred = 673.3147099767981
ytrue = bank_test['Credit Score']


# Calculate errors for the baseline model using the average of credit scores in testing set
y_bar = bank_test['Credit Score'].mean()

# Compute MAE_Baseline

def mae(ytrue): #Creating a function to calculate MAE_baseline
	sum_error = 0.0
	for i in range(len(ytrue)):
		sum_error += abs(ytrue[i] - y_bar)
	return sum_error / float(len(ytrue))

ytrue = bank_test['Credit Score']
MAE_baseline = mae(ytrue)


### The MAE_Regreession ### (obtained by passing the test data through the model developed by the training set)

ypred_R = RegModel.predict(X_test) # Obtain Predicted Values
ytrue_R = bank_test['Credit Score'] # Obtain True Values

MAE_Regression = mean_absolute_error(ytrue_R, ypred_R) # Calculating MAE

# Print both MAE Scores
print('The MAE for the Baseline Model is equal to ', MAE_baseline)
print('The MAE for the Regression Model is equal to ', MAE_Regression)

The MAE for the Baseline Model is equal to  48.60024069638013
The MAE for the Regression Model is equal to  47.79066993785576


Since 47.79 = MAE_Regression < MAE_Baseline = 48.60, the regression model outperformed the baseline model. 